### PR TITLE
Make GCS multipart upload failures S3-compatible

### LINKED ISSUE #206: aws-sdk-php `testMultipartUploadFailure` fails on GCS gateway
The test fails on GCS gateway since part size validation is missing.

```
Running aws-sdk-php tests ... FAILED in 49 seconds
{
  "name": "aws-sdk-php",
  "function": "uploadPart ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-bucket-14203",
    "Object": "obj1"
  },
  "duration": "6079",
  "status": "FAIL",
  "error": "Expected completeMultipartUpload to fail with EntityTooSmall"
}
```

### PR DESCRIPTION
**Pull request by krisis**
## Description

This change translates GCS error responses from multipart upload failures to appropriate AWS S3 compatible errors.

## Motivation and Context

Fixes minio/mint#206

## How Has This Been Tested?

Running mint aws-sdk-php tests locally

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by balamurugana — APPROVED** [review]
LGTM

File: cmd/gateway-gcs.go [hunk 1]
```diff
@@ -1036,9 +1046,26 @@ func (l *gcsGateway) CompleteMultipartUpload(bucket string, key string, uploadID
 	}
 
 	var parts []*storage.ObjectHandle
-	for _, uploadedPart := range uploadedParts {
+	partSizes := make([]int64, len(uploadedParts))
+	for i, uploadedPart := range uploadedParts {
 		parts = append(parts, l.client.Bucket(bucket).Object(gcsMultipartDataName(uploadID,
 			uploadedPart.PartNumber, uploadedPart.ETag)))
+		partAttr, pErr := l.client.Bucket(bucket).Object(gcsMultipartDataName(uploadID, uploadedPart.PartNumber, uploadedPart.ETag)).Attrs(l.ctx)
```
Comment (krisis) [reply cmd/gateway-gcs.go:1053]: > Object returns an ObjectHandle, which provides operations on the named object. This call does not perform any network operations.

A network call to fetch object attributes is made when `ObjectHandle.Attrs` method is called. So, there is only one call to the backend here.

File: cmd/gateway-gcs.go [hunk 2]
```diff
@@ -192,12 +192,16 @@ func gcsToObjectError(err error, params ...string) error {
 
 	bucket := ""
 	object := ""
+	uploadID := ""
 	if len(params) >= 1 {
 		bucket = params[0]
 	}
 	if len(params) == 2 {
 		object = params[1]
```
Comment (krisis) [reply cmd/gateway-gcs.go:200]: `bucket` and `object` values aren't used when we return error of type `InvalidUpload`, which is the only place where we use `uploadID`. This is intentionally left as is.

Comment (donatello) [inline cmd/gateway-gcs.go:200]: This `if` block needs to be changed to `if len(params) >=2` or `object` will be empty when len(params) == 3.

File: cmd/gateway-gcs.go (hunk 1 above)
Comment (donatello) [inline cmd/gateway-gcs.go:1053]: Looks like there is an extra backend call here - the `i`-th part is being fetched twice (before the call to `.Attrs()`). The result from the first call should be reused.

**Review by donatello — CHANGES_REQUESTED** [review]

**Review by donatello — APPROVED** [review]

**Comment by deekoder** [comment]
ignoring codecov as spurious failure.

**Pull request merged**